Display driver API sample

vanwinkeljan opened the pull request:
This PR introduces a generic display shield sample as long as a board supports arduino shields this sample should work for suche board.

vanwinkeljan committed:
samples: display: Added generic display shield sample

Added a generic display shield sample.

Signed-off-by: Jan Van Winkel <jan.van_winkel@dxplore.eu>

jfischer-phytec-iot commented on boards/shields/adafruit_2_8_tft_touch_v2/dts_fixup.h:
```diff
@@ -0,0 +1 @@
+#define DT_DISPLAY_SHIELD_DEV_NAME DT_INST_0_ILITEK_ILI9340_LABEL
```
I suggest to use CONFIG_LVGL_DISPLAY_DEV_NAME as it is already there.

vanwinkeljan replied:
That is not an option `CONFIG_LVGL_DISPLAY_DEV_NAME` can only be set if LVGL is used else this config is not available.

jfischer-phytec-iot commented:
@vanwinkeljan Why not make it more generic and move to samples/drivers/? I guess your plan is also to remove `samples/display/ili9340` and `samples/display/st7789`?

dbkinder commented on samples/display/display_shield/README.rst:
```diff
@@ -0,0 +1,46 @@
+The rectangle colors and positions are chosen so that you can check the
+orientation of the LCD and correct RGB bit order. The rectangles are drawn
+in clockwise order, from top left corner: Red, Green, Blue, grey. The shade of
+grey changes from black through to white. (if the grey looks too green or red
```
change ``if`` to ``If`` (beginning of a sentence)

dbkinder commented on samples/display/display_shield/README.rst:
```diff
@@ -0,0 +1,46 @@
+Building and Running
+********************
+
+This sample makes use of arduino based display shields and should work on any
```
Let's fix some issues with this rewrite:
```
This sample makes use of Arduino-based display shields and should work on any
board supporting such a shield, for example:

- :ref:`adafruit_2_8_tft_touch_v2`
- :ref:`ssd1306_128x32_shield`
- :ref:`ssd1306_128x64_shield`
```

dbkinder commented on samples/display/display_shield/README.rst:
```diff
@@ -0,0 +1,46 @@
+   :compact:
+
+For testing purpose without the need of any hardware, the :ref:`native_posix`
+board is also supported and can be build as follows;
```
change ``build`` to ``built``

vanwinkeljan commented:
> @vanwinkeljan Why not make it more generic and move to samples/drivers/? I guess your plan is also to remove samples/display/ili9340 and samples/display/st7789?

@jfischer-phytec-iot  thx for the suggestion will take care of this in the next push.

erwango reviewed (changes requested):
This addition is not matching the definition of shield: a HW component that could be added on top of a board thanks to compatible connector.
Agree with @jfischer-phytec-iot , it would be better proposed as a sample/driver.

vanwinkeljan commented:
@erwango The sample was already moved to `samples/drivers/display_shield`

MaureenHelm reviewed (changes requested):
> > @vanwinkeljan Why not make it more generic and move to samples/drivers/? I guess your plan is also to remove samples/display/ili9340 and samples/display/st7789?
> 
> @jfischer-phytec-iot thx for the suggestion will take care of this in the next push.

This looks more like a general display driver sample than a shield sample, so I think it belongs in `samples/drivers/display`. It should work on mimxrt1050_evk, which doesn't need a shield.

vanwinkeljan commented:
@MaureenHelm Valid point, only remaining problem is how do we get the correct display name in the sample.

An option would to add a `dts_fixup.h` file to the sample to map all display name defines to a well known define that we can use in the sample.

jfischer-phytec-iot commented:
>Valid point, only remaining problem is how do we get the correct display name in the sample.

I am to implement similar mechanism that network stack has to discover the network interfaces or what we have for the usb classes.

vanwinkeljan commented:
@jfischer-phytec-iot Any suggestion how we proceed with this PR, eg. proceed with a dts_fixup.h file until your mechanism is in place, ... ?

MaureenHelm commented:
> only remaining problem is how do we get the correct display name in the sample.

What about a common dts alias like we do for led0?

jfischer-phytec-iot commented:
> > only remaining problem is how do we get the correct display name in the sample.
> 
> What about a common dts alias like we do for led0?

Would probably not work for the configuration like reel board + ssd1306_128x32 or mimxrt1050 + waveshare_epaper.

jfischer-phytec-iot commented:
> @jfischer-phytec-iot Any suggestion how we proceed with this PR, eg. proceed with a dts_fixup.h file until your mechanism is in place, ... ?

I think yes if that with dts_fixup.h works and is acceptable.

dbkinder commented on samples/drivers/display_shield/README.rst:
```diff
@@ -0,0 +1,44 @@
+- :ref:`ssd1306_128x32_shield`
+- :ref:`ssd1306_128x64_shield`
+
+Bellow is an example on how to build for a :ref:`nrf52840_pca10056` board with a
```
Change to ``Below``

dbkinder commented on samples/drivers/display_shield/README.rst:
```diff
@@ -0,0 +1,44 @@
+This sample will draw some basic rectangles onto the display.
+The rectangle colors and positions are chosen so that you can check the
+orientation of the LCD and correct RGB bit order. The rectangles are drawn
+in clockwise order, from top left corner: Red, Green, Blue, grey. The shade of
```
for consistency, use lowercase for the color names (red, green, blue)

dbkinder commented on samples/drivers/display_shield/README.rst:
```diff
@@ -0,0 +1,44 @@
+orientation of the LCD and correct RGB bit order. The rectangles are drawn
+in clockwise order, from top left corner: Red, Green, Blue, grey. The shade of
+grey changes from black through to white. If the grey looks too green or red
+at any point then the LCD may be endian swapped.
```
What should you do if the LCD is "endian swapped"?  Is there a configuration or code change needed?

vanwinkeljan replied:
As this is a generic display sample there is no clear solution to this issue and it is up to the user to figure it out. Could be a Kconfig that is not set correctly or a plain bug in the code.

vanwinkeljan commented:
Updated PR
 * Renamed sample to `display` instead of `display_shield`
 * Added support for `mimxrt1050_evk` (including build in CI)
 * Run sample in CI for `native_posix` + dummy display
 * Added `waveshare_epaper_gdeh0213b1` to CI
 * Removed shield specific `dts_fixup.h`
 * Added `dts_fixup.h` to sample until display discovery mechanism is available

vanwinkeljan commented:
Update: Added myself to `CODEOWNERS` for sample

MaureenHelm commented on samples/drivers/display/boards/mimxrt1050_evk.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_DISPLAY_MCUX_ELCDIF=y
```
This shouldn't be necessary because of:
https://github.com/zephyrproject-rtos/zephyr/blob/d58f8bee79dddb9aa1b65a2ef1624ebf1b389a4e/soc/arm/nxp_imx/rt/Kconfig.defconfig.series#L28-L33

vanwinkeljan replied:
removed

MaureenHelm commented on samples/drivers/display/README.rst:
```diff
@@ -0,0 +1,47 @@
+orientation of the LCD and correct RGB bit order. The rectangles are drawn
+in clockwise order, from top left corner: red, green, blue, grey. The shade of
+grey changes from black through to white. If the grey looks too green or red
+at any point then the LCD may be endian swapped.
```
I had to byte swap `fill_buffer_rgb565` to get it to look right on mimxrt1050_evk

vanwinkeljan replied:
Sound like an endian issue mcux_elcdif driver, we could add bgr565 as pixel format, update the API header file to specify that the pixel formats are byte-order and return bgr565 from the mcux_elcdif driver.

dbkinder replied:
That's funny.  My previous comments ask what someone should do if indeed the LCD is endian swapped and @MaureenHelm hit this.  So maybe this is a common enough problem to say more here about what the developer might need to do in this situation?

vanwinkeljan replied:
@dbkinder Problem is that there is not a common solution and it just a bug in the display driver.

MaureenHelm replied:
I'm not convinced it's a bug in the driver. If that were the case, wouldn't lvgl also not work? It looks like the sample is assuming a particular endianness with:
```c
	for (size_t idx = 0; idx < buf_size; idx += 2) {
		*(buf + idx + 0) = (color >> 8) & 0xFFu;
		*(buf + idx + 1) = (color >> 0) & 0xFFu;
	}
```

vanwinkeljan replied:
> wouldn't lvgl also not work?

Yes if they would use RGB565 byte ordering but their code uses BGR565

> It looks like the sample is assuming a particular endianness

Actually the API is assuming this, if the code is changed it will break the SPI based display shields

MaureenHelm replied:
Agreed. Per your earlier suggestion, I think we need to add BGR565 as a pixel format and clarify the byte order in the API.

vanwinkeljan replied:
@MaureenHelm Ok I will take care of it, is it ok I do this in a separate PR?

MaureenHelm replied:
Yes, thank you

dbkinder commented on samples/drivers/display/README.rst:
```diff
@@ -0,0 +1,47 @@
+Building and Running
+********************
+
+As this is a generic sample it should work with any display supported by zephyr.
```
capitalize ``Zephyr``

dbkinder reviewed (approved):
docs LGTM, thanks.

vanwinkeljan commented:
@MaureenHelm @erwango could you please revisit your reviews?

erwango commented:
@vanwinkeljan,approved.
Btw, can you provide inputs on https://github.com/zephyrproject-rtos/zephyr/pull/20934.
I've modified shield Kconfig system, so I had to rewrite some shields configuration. 
I have some doubts on the display shields configuration I proposed (particularly where to draw the line between application and shield configuration), and it seems like this change is perfect to test the proposal.

vanwinkeljan commented:
@erwango I will have a look at the other PR

jfischer-no reviewed (changes requested):
block it to have time for review and test

jfischer-phytec-iot commented on samples/drivers/display/src/main.c:
```diff
@@ -0,0 +1,272 @@
+	buf_desc.height = h_step;
+
+	for (int idx = 0; idx < capabilities.y_resolution; idx += h_step) {
+		display_write(display_dev, 0, idx, &buf_desc, buf);
```
I have tested it with ssd1306, reel board, and adafruit_2_8_tft_touch_v2 for a while. Monochrome displays do not work with this sample if the h_step is not a multiple of an octet, there are also few issues in the drivers. I got it working as intended on ssd1306 with the changes below. I also suggest to define rect_w and rect_h as minimally addressable area for a display type and the scale it up. Also I observed that this sample did not work on ili9340 display on the Adafruit shield until I flashed it with samples/gui/lvgl first, maybe an initialization issue in the driver self.

```diff
diff --git a/samples/drivers/display/src/main.c b/samples/drivers/display/src/main.c
index b74c3c557a..ea7481e321 100644
--- a/samples/drivers/display/src/main.c
+++ b/samples/drivers/display/src/main.c
@@ -154,7 +154,7 @@ void main(void)
        size_t y;
        size_t rect_w;
        size_t rect_h;
-       size_t h_step;
+       size_t h_step = 0;
        size_t grey_count;
        u8_t *buf;
        s32_t grey_scale_sleep;
@@ -169,7 +169,8 @@ void main(void)
        display_dev = device_get_binding(DISPLAY_DEV_NAME);
 
        if (display_dev == NULL) {
-               LOG_ERR("Device not found. Aborting sample.");
+               LOG_ERR("Device %s not found. Aborting sample.",
+                       DISPLAY_DEV_NAME);
                RETURN_FROM_MAIN(1);
        }
 
@@ -193,19 +194,23 @@ void main(void)
        case PIXEL_FORMAT_ARGB_8888:
                fill_buffer_fnc = fill_buffer_argb8888;
                buf_size = 4 * rect_w * rect_h;
+               h_step = (rect_w * rect_h) / capabilities.x_resolution;
                break;
        case PIXEL_FORMAT_RGB_888:
                fill_buffer_fnc = fill_buffer_rgb888;
                buf_size = 3 * rect_w * rect_h;
+               h_step = (rect_w * rect_h) / capabilities.x_resolution;
                break;
        case PIXEL_FORMAT_RGB_565:
                fill_buffer_fnc = fill_buffer_rgb565;
                buf_size = 2 * rect_w * rect_h;
+               h_step = (rect_w * rect_h) / capabilities.x_resolution;
                break;
        case PIXEL_FORMAT_MONO01:
        case PIXEL_FORMAT_MONO10:
                fill_buffer_fnc = fill_buffer_mono;
-               buf_size = rect_w * rect_h / 8;
+               buf_size = capabilities.x_resolution * rect_h / 8;
+               h_step = rect_h;
                break;
        default:
                LOG_ERR("Unsupported pixel format. Aborting sample.");
@@ -220,7 +225,6 @@ void main(void)
        }
 
        (void)memset(buf, 0xFFu, buf_size);
-       h_step = (rect_w * rect_h) / capabilities.x_resolution;
 
        buf_desc.buf_size = buf_size;
        buf_desc.pitch = capabilities.x_resolution;
```

vanwinkeljan replied:
@jfischer-phytec-iot thx for testing!

>  I got it working as intended on ssd1306 with the changes below.

Will take the changes along

>  I also suggest to define rect_w and rect_h as minimally addressable area for a display type and the scale it up. 

I will update the code

> Also I observed that this sample did not work on ili9340 display on the Adafruit shield until I flashed it with samples/gui/lvgl first, maybe an initialization issue in the driver self.

Hmm strange I also tested in a `adafruit_2_8_tft_touch_v2` shield and did not see any issue.
I will repeat the test and see what comes out of it.

vanwinkeljan replied:
@jfischer-phytec-iot updated the code

On `adafruit_2_8_tft_touch_v2` shield the sample works for me out of the box* with the exception that the top left pixels of the rectangle are not correct

_*_ Tested with a `nrf52840_pca10056` preview board so I had to update the dts to use SPI 1 instead of 3

jfischer-phytec-iot commented on samples/drivers/display/dts_fixup.h:
```diff
@@ -0,0 +1,30 @@
+#ifdef DT_INST_0_FSL_IMX6SX_LCDIF_LABEL
+#define DISPLAY_DEV_NAME DT_INST_0_FSL_IMX6SX_LCDIF_LABEL
+#endif
+
```
please add

```diff
diff --git a/samples/drivers/display/dts_fixup.h b/samples/drivers/display/dts_fixup.h
index 46fb3a9d1e..dd69012cb4 100644
--- a/samples/drivers/display/dts_fixup.h
+++ b/samples/drivers/display/dts_fixup.h
@@ -20,6 +20,10 @@
 #define DISPLAY_DEV_NAME DT_INST_0_FSL_IMX6SX_LCDIF_LABEL
 #endif
 
+#ifdef DT_INST_0_SITRONIX_ST7789V_LABEL
+#define DISPLAY_DEV_NAME DT_INST_0_SITRONIX_ST7789V_LABEL
+#endif
+
 #ifdef CONFIG_SDL_DISPLAY_DEV_NAME
 #define DISPLAY_DEV_NAME CONFIG_SDL_DISPLAY_DEV_NAME
 #endif
```

vanwinkeljan replied:
done

jfischer-phytec-iot replied:
I guess you forgot git add ..., ?

vanwinkeljan replied:
worse I forgot to commit, pushed now

jfischer-phytec-iot commented on samples/drivers/display/sample.yaml:
```diff
@@ -0,0 +1,34 @@
+  sample.display.shield.waveshare_epaper_gdeh0213b1:
+    platform_whitelist: nrf52840_pca10056
+    extra_args: SHIELD=waveshare_epaper_gdeh0213b1
+    tags: display shield
```
please add
```diff
diff --git a/samples/drivers/display/sample.yaml b/samples/drivers/display/sample.yaml
index b709f1a6b3..fcb6af8b44 100644
--- a/samples/drivers/display/sample.yaml
+++ b/samples/drivers/display/sample.yaml
@@ -18,6 +18,16 @@ tests:
     platform_whitelist: nrf52840_pca10056
     extra_args: SHIELD=waveshare_epaper_gdeh0213b1
     tags: display shield
+  sample.display.shield.st7789v_tl019fqv01:
+    build_only: true
+    platform_whitelist: nrf52_pca10040
+    extra_args: SHIELD=st7789v_tl019fqv01
+    tags: display shield
+  sample.display.shield.st7789v_waveshare_240x240:
+    build_only: true
+    platform_whitelist: nrf52_pca10040
+    extra_args: SHIELD=st7789v_waveshare_240x240
+    tags: display shield
   sample.display.mcux_elcdif:
     platform_whitelist: mimxrt1050_evk
     tags: display

```

vanwinkeljan replied:
done

jfischer-no reviewed (approved):
LGTM, thanks. Tested with ssd1306, mimxrt1050, st7789v, ssd16xx. As next we should remove samples/display/ili9340 and sample/display/st7789v.

jfischer-phytec-iot commented on samples/drivers/display/README.rst:
```diff
@@ -0,0 +1,47 @@
+*************************************
+
+- :ref:`adafruit_2_8_tft_touch_v2`
+- :ref:`ssd1306_128x32_shield`
```
This has to be fixed to pass Documentation checks:

```diff
diff --git a/samples/drivers/display/README.rst b/samples/drivers/display/README.rst
index 99a66aa2a3..731de70ca2 100644
--- a/samples/drivers/display/README.rst
+++ b/samples/drivers/display/README.rst
@@ -42,6 +42,5 @@ List of Arduino-based display shields
 *************************************
 
 - :ref:`adafruit_2_8_tft_touch_v2`
-- :ref:`ssd1306_128x32_shield`
-- :ref:`ssd1306_128x64_shield`
+- :ref:`ssd1306_128_shield`
 - :ref:`waveshare_e_paper_raw_panel_shield`
```

vanwinkeljan replied:
thx for the input, doc has been updated

jfischer-phytec-iot commented:
@MaureenHelm Please revise.
